### PR TITLE
Change getService() to throw error when service does not exist

### LINKED ISSUE #89: Make T3 throw or warn when attempting to retrieve a service that doesn't exist in debug mode
Recently we discovered a ~6mo old bug in our system that was able to persist because T3 silently allowed a features flip service that didn't exist to be retrieved and used as a falsey value in subsequent logic. This could have been identified if T3 had thrown upon attempting to fetch the service, or at least dumped to conosle.warn when in debug mode to at least let a developer know they're attempting to access a service that does not exist.

For code cleanliness console statements can then later be stripped out during our build process with the drop_console flag when the code is minified for production use.

### PR DESCRIPTION
**Pull request by j3tan**
- Added hasService() method to allow for optional services
- debug off = fires an error and returns null

fixes #89

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @j3tan has signed the CLA. Thanks for the pull request!

**Comment by j3tan** [comment]
@jdivock @nzakas - please check this out to see if this works as you would expect it to.

File: lib/application-stub.js [hunk 1]
```diff
@@ -69,6 +69,15 @@
 			},
 
 			/**
+			 * Checks if a service exists
+			 * @param {string} serviceName The name of the service to check.
+			 * @returns {boolean} True, if service exist. False, otherwise.
+			 */
+			hasService: function(serviceName) {
+				return serviceName in services;
```
Comment (nzakas) [inline lib/application-stub.js]: I'd switch to `hasOwnProperty()` here. Right now, still will return `true` when `serviceName` is `toString`, `hasOwnProperty`, etc.

File: lib/application.js [hunk 2]
```diff
@@ -693,6 +701,14 @@ Box.Application = (function() {
 		 */
 		getService: getService,
 
+		/**
+		 * Checks if a service exists
+		 * @param {string} serviceName The name of the service to check.
+		 * @returns {boolean} True, if service exist. False, otherwise.
+		 */
+		hasService: function(serviceName) {
+			return serviceName in services;
```
Comment (nzakas) [inline lib/application.js]: Ditto

File: lib/test-service-provider.js [hunk 3]
```diff
@@ -65,12 +68,23 @@
...
 			return null;
 		},
 
 		/**
+		 * Checks if a service exists
+		 * @param {string} serviceName The name of the service to check.
+		 * @returns {boolean} True, if service exist. False, otherwise.
+		 */
+		hasService: function(serviceName) {
+			return serviceName in this.stubs || application.hasService(serviceName);
```
Comment (nzakas) [inline lib/test-service-provider.js]: Ditto

File: tests/application-test.js [hunk 4]
```diff
@@ -530,6 +550,19 @@ describe('Box.Application', function() {
...
 		});
 
+		describe('hasService()', function() {
+
+			it('should return true when the service exists', function() {
+				Box.Application.addService('test', function() {});
+				assert.isTrue(Box.Application.hasService('test'));
+			});
+
+			it('should return false when the service does not exist', function() {
```
Comment (nzakas) [inline tests/application-test.js]: Add test for "toString" to see the problem.

Comment (j3tan) [inline tests/application-test.js]: tested and now it works

File: lib/application-stub.js (hunk 1 above)
Comment (j3tan) [inline lib/application-stub.js]: done for all

**Comment by nzakas** [comment]
+2

**Pull request merged**